### PR TITLE
feat: MirroringConnection: timeout in close() and abort()

### PR DESCRIPTION
**Pull request by prawilny**
This change is [Reviewable](https://reviewable.io/reviews/unoperate/java-bigtable-hbase-1/133)

TestBlocking#testConnectionCloseBlocksUntilAllRequestsHaveBeenVerified doesn't pass - also on master.

### CODE REVIEW

**Review by mwalkiewicz — CHANGES_REQUESTED** [review]
> Reviewed 3 of 4 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/unoperate/java-bigtable-hbase-1/133)* status: 3 of 4 files reviewed, 6 unresolved discussions (waiting on @mwalkiewicz and @prawilny)

___
*[bigtable-hbase-mirroring-client-1.x-parent/bigtable-hbase-mirroring-client-1.x/src/main/java/com/google/cloud/bigtable/mirroring/hbase1_x/MirroringConnection.java, line 283 at r1](https://reviewable.io/reviews/unoperate/java-bigtable-hbase-1/133#-Mnl9rtn2hmEaAW7AZMm:-Mnl9rtn2hmEaAW7AZMn:b-lzq3ew) ([raw file](https://github.com/unoperate/java-bigtable-hbase-1/blob/cf1d870127b5ebf4689067116c4140b1387731df/bigtable-hbase-mirroring-client-1.x-parent/bigtable-hbase-mirroring-client-1.x/src/main/java/com/google/cloud/bigtable/mirroring/hbase1_x/MirroringConnection.java#L283)):*
> ```Java
> 
>   private void terminateConnectionsWithTimeout(
>       final Callable<Void> terminatingAction, boolean throwExceptionIfErred) throws IOException {
> ```

There is something called CallableThrowingIOException in our project, we can use it here to make exceptions thrown by terminatingAction more specific.
___
*[bigtable-hbase-mirroring-client-1.x-parent/bigtable-hbase-mirroring-client-1.x/src/main/java/com/google/cloud/bigtable/mirroring/hbase1_x/MirroringConnection.java, line 295 at r1](https://reviewable.io/reviews/unoperate/java-bigtable-hbase-1/133#-Mnl9faE9Iv1V66tNQC3:-Mnl9faE9Iv1V66tNQC4:bcvjrwc) ([raw file](https://github.com/unoperate/java-bigtable-hbase-1/blob/cf1d870127b5ebf4689067116c4140b1387731df/bigtable-hbase-mirroring-client-1.x-parent/bigtable-hbase-mirroring-client-1.x/src/main/java/com/google/cloud/bigtable/mirroring/hbase1_x/MirroringConnection.java#L295)):*
> ```Java
>       } catch (Exception e) {
>         if (throwExceptionIfErred) {
>           throw new IOException(e);
> ```

Can't we throw this exception unconditionally and just catch and ignore it in abort? We are already ignoring something there.
___
*[bigtable-hbase-mirroring-client-1.x-parent/bigtable-hbase-mirroring-client-1.x/src/main/java/com/google/cloud/bigtable/mirroring/hbase1_x/MirroringConnection.java, line 301 at r1](https://reviewable.io/reviews/unoperate/java-bigtable-hbase-1/133#-Mnl75GS8ki6S8pLA8ZS:-Mnl75GT8PkgQEzswZPD:b-e729yn) ([raw file](https://github.com/unoperate/java-bigtable-hbase-1/blob/cf1d870127b5ebf4689067116c4140b1387731df/bigtable-hbase-mirroring-client-1.x-parent/bigtable-hbase-mirroring-client-1.x/src/main/java/com/google/cloud/bigtable/mirroring/hbase1_x/MirroringConnection.java#L301)):*
> ```Java
>       assert false;
>     } catch (InterruptedException e) {
>       throw new RuntimeException(e);
> ```

Why won't we throw a `IOException(e)`?
___
*[bigtable-hbase-mirroring-client-1.x-parent/bigtable-hbase-mirroring-client-1.x/src/main/java/com/google/cloud/bigtable/mirroring/hbase1_x/MirroringConnection.java, line 302 at r1](https://reviewable.io/reviews/unoperate/java-bigtable-hbase-1/133#-Mnl7mbW9EAJv7kXFvZ4:-Mnl7mbX34YhjQ7kk-8W:bokpmxe) ([raw file](https://github.com/unoperate/java-bigtable-hbase-1/blob/cf1d870127b5ebf4689067116c4140b1387731df/bigtable-hbase-mirroring-client-1.x-parent/bigtable-hbase-mirroring-client-1.x/src/main/java/com/google/cloud/bigtable/mirroring/hbase1_x/MirroringConnection.java#L302)):*
> ```Java
>     } catch (InterruptedException e) {
>       throw new RuntimeException(e);
>     } catch (TimeoutException e) {
> ```

I think that we should an error to a logger (to stdout) here.
___
*[bigtable-hbase-mirroring-client-1.x-parent/bigtable-hbase-mirroring-client-1.x/src/main/java/com/google/cloud/bigtable/mirroring/hbase1_x/MirroringConnection.java, line 302 at r1](https://reviewable.io/reviews/unoperate/java-bigtable-hbase-1/133#-Mnl8-lQ0v9pXR2yA6xZ:-Mnl8-lQ0v9pXR2yA6x_:bs63m5a) ([raw file](https://github.com/unoperate/java-bigtable-hbase-1/blob/cf1d870127b5ebf4689067116c4140b1387731df/bigtable-hbase-mirroring-client-1.x-parent/bigtable-hbase-mirroring-client-1.x/src/main/java/com/google/cloud/bigtable/mirroring/hbase1_x/MirroringConnection.java#L302)):*
> ```Java
>     } catch (InterruptedException e) {
>       throw new RuntimeException(e);
>     } catch (TimeoutException e) {
> ```

So, if the timeout happens, the connections are not closed and the user is not notified. I think that we can throw a IOException(e) after scheduling the terminatingAction and let the user know that there is something went wrong (because the primary was not closed).
___
*[bigtable-hbase-mirroring-client-1.x-parent/bigtable-hbase-mirroring-client-1.x/src/main/java/com/google/cloud/bigtable/mirroring/hbase1_x/utils/MirroringConfigurationHelper.java, line 136 at r1](https://reviewable.io/reviews/unoperate/java-bigtable-hbase-1/133#-Mnj2GIf1oYYnSiM1PFi:-Mnj2GIf1oYYnSiM1PFj:b53v1kh) ([raw file](https://github.com/unoperate/java-bigtable-hbase-1/blob/cf1d870127b5ebf4689067116c4140b1387731df/bigtable-hbase-mirroring-client-1.x-parent/bigtable-hbase-mirroring-client-1.x/src/main/java/com/google/cloud/bigtable/mirroring/hbase1_x/utils/MirroringConfigurationHelper.java#L136)):*
> ```Java
>    * pending operations before terminating connections with primary and secondary databases.
>    *
>    * <p>Defaults to 1000. If the wait times out, the termination is scheduled asynchronously.
> ```

I'd set the default to 60 seconds, we really want to wait for all the writes to finish.

**Review by prawilny — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/unoperate/java-bigtable-hbase-1/133)* status: 1 of 4 files reviewed, 6 unresolved discussions (waiting on @mwalkiewicz)

___
*[bigtable-hbase-mirroring-client-1.x-parent/bigtable-hbase-mirroring-client-1.x/src/main/java/com/google/cloud/bigtable/mirroring/hbase1_x/MirroringConnection.java, line 283 at r1](https://reviewable.io/reviews/unoperate/java-bigtable-hbase-1/133#-Mnl9rtn2hmEaAW7AZMm:-Mo538ojCOPkgcxa9czG:b-896fix) ([raw file](https://github.com/unoperate/java-bigtable-hbase-1/blob/cf1d870127b5ebf4689067116c4140b1387731df/bigtable-hbase-mirroring-client-1.x-parent/bigtable-hbase-mirroring-client-1.x/src/main/java/com/google/cloud/bigtable/mirroring/hbase1_x/MirroringConnection.java#L283)):*
<details><summary><i>Previously, mwalkiewicz (Mateusz Walkiewicz) wrote…</i></summary><blockquote>

There is something called CallableThrowingIOException in our project, we can use it here to make exceptions thrown by terminatingAction more specific.
</blockquote></details>

Done.
___
*[bigtable-hbase-mirroring-client-1.x-parent/bigtable-hbase-mirroring-client-1.x/src/main/java/com/google/cloud/bigtable/mirroring/hbase1_x/MirroringConnection.java, line 295 at r1](https://reviewable.io/reviews/unoperate/java-bigtable-hbase-1/133#-Mnl9faE9Iv1V66tNQC3:-Mo55e7m6irNExapu6q9:b-l6tdiy) ([raw file](https://github.com/unoperate/java-bigtable-hbase-1/blob/cf1d870127b5ebf4689067116c4140b1387731df/bigtable-hbase-mirroring-client-1.x-parent/bigtable-hbase-mirroring-client-1.x/src/main/java/com/google/cloud/bigtable/mirroring/hbase1_x/MirroringConnection.java#L295)):*
<details><summary><i>Previously, mwalkiewicz (Mateusz Walkiewicz) wrote…</i></summary><blockquote>

Can't we throw this exception unconditionally and just catch and ignore it in abort? We are already ignoring something there.
</blockquote></details>

Yes, we can. Done.
___
*[bigtable-hbase-mirroring-client-1.x-parent/bigtable-hbase-mirroring-client-1.x/src/main/java/com/google/cloud/bigtable/mirroring/hbase1_x/MirroringConnection.java, line 301 at r1](https://reviewable.io/reviews/unoperate/java-bigtable-hbase-1/133#-Mnl75GS8ki6S8pLA8ZS:-Mo55coU2PSI2iodQ9oG:bbyytu0) ([raw file](https://github.com/unoperate/java-bigtable-hbase-1/blob/cf1d870127b5ebf4689067116c4140b1387731df/bigtable-hbase-mirroring-client-1.x-parent/bigtable-hbase-mirroring-client-1.x/src/main/java/com/google/cloud/bigtable/mirroring/hbase1_x/MirroringConnection.java#L301)):*
<details><summary><i>Previously, mwalkiewicz (Mateusz Walkiewicz) wrote…</i></summary><blockquote>

Why won't we throw a `IOException(e)`?
</blockquote></details>

Yep, it's better.
___
*[bigtable-hbase-mirroring-client-1.x-parent/bigtable-hbase-mirroring-client-1.x/src/main/java/com/google/cloud/bigtable/mirroring/hbase1_x/MirroringConnection.java, line 302 at r1](https://reviewable.io/reviews/unoperate/java-bigtable-hbase-1/133#-Mnl7mbW9EAJv7kXFvZ4:-Mo54U0PCEmG5ZTHNQhA:b-896fix) ([raw file](https://github.com/unoperate/java-bigtable-hbase-1/blob/cf1d870127b5ebf4689067116c4140b1387731df/bigtable-hbase-mirroring-client-1.x-parent/bigtable-hbase-mirroring-client-1.x/src/main/java/com/google/cloud/bigtable/mirroring/hbase1_x/MirroringConnection.java#L302)):*
<details><summary><i>Previously, mwalkiewicz (Mateusz Walkiewicz) wrote…</i></summary><blockquote>

I think that we should an error to a logger (to stdout) here.
</blockquote></details>

Done.
___
*[bigtable-hbase-mirroring-client-1.x-parent/bigtable-hbase-mirroring-client-1.x/src/main/java/com/google/cloud/bigtable/mirroring/hbase1_x/MirroringConnection.java, line 302 at r1](https://reviewable.io/reviews/unoperate/java-bigtable-hbase-1/133#-Mnl8-lQ0v9pXR2yA6xZ:-Mo54loG-ODh4KEwLvYG:b-896fix) ([raw file](https://github.com/unoperate/java-bigtable-hbase-1/blob/cf1d870127b5ebf4689067116c4140b1387731df/bigtable-hbase-mirroring-client-1.x-parent/bigtable-hbase-mirroring-client-1.x/src/main/java/com/google/cloud/bigtable/mirroring/hbase1_x/MirroringConnection.java#L302)):*
<details><summary><i>Previously, mwalkiewicz (Mateusz Walkiewicz) wrote…</i></summary><blockquote>

So, if the timeout happens, the connections are not closed and the user is not notified. I think that we can throw a IOException(e) after scheduling the terminatingAction and let the user know that there is something went wrong (because the primary was not closed).
</blockquote></details>

Done.
___
*[bigtable-hbase-mirroring-client-1.x-parent/bigtable-hbase-mirroring-client-1.x/src/main/java/com/google/cloud/bigtable/mirroring/hbase1_x/utils/MirroringConfigurationHelper.java, line 136 at r1](https://reviewable.io/reviews/unoperate/java-bigtable-hbase-1/133#-Mnj2GIf1oYYnSiM1PFi:-Mo54zOt6nX0TFX8NbP5:b-896fix) ([raw file](https://github.com/unoperate/java-bigtable-hbase-1/blob/cf1d870127b5ebf4689067116c4140b1387731df/bigtable-hbase-mirroring-client-1.x-parent/bigtable-hbase-mirroring-client-1.x/src/main/java/com/google/cloud/bigtable/mirroring/hbase1_x/utils/MirroringConfigurationHelper.java#L136)):*
<details><summary><i>Previously, mwalkiewicz (Mateusz Walkiewicz) wrote…</i></summary><blockquote>

I'd set the default to 60 seconds, we really want to wait for all the writes to finish.
</blockquote></details>

Done.

**Review by prawilny — COMMENTED** [review]
Thinking about https://github.com/googleapis/java-bigtable-hbase/pull/3322#discussion_r748596122 lead me to conclusion that abort() should have a simpler implementation and not wait for background requests.

> *[Reviewable](https://reviewable.io/reviews/unoperate/java-bigtable-hbase-1/133#-:-ModmKJe685oVe-SmG1w:bm6fn3n)* status: 0 of 4 files reviewed, 6 unresolved discussions (waiting on @mwalkiewicz)

File: bigtable-hbase-mirroring-client-1.x-parent/bigtable-hbase-mirroring-client-1.x/src/main/java/com/google/cloud/bigtable/mirroring/hbase1_x/MirroringConnection.java [hunk 1]
```diff
@@ -277,12 +283,36 @@ public Connection getSecondaryConnection() {
...
-      this.referenceCounter.getOnLastReferenceClosed().get();
+      this.referenceCounter
+          .getOnLastReferenceClosed()
+          .get(
+              this.configuration.mirroringOptions.connectionTerminationTimeoutMillis,
+              TimeUnit.MILLISECONDS);
+      terminatingAction.call();
     } catch (ExecutionException e) {
-      throw new RuntimeException(e);
+      assert false;
```
Comment (mwalkiewicz) [inline bigtable-hbase-mirroring-client-1.x-parent/bigtable-hbase-mirroring-client-1.x/src/main/java/com/google/cloud/bigtable/mirroring/hbase1_x/MirroringConnection.java:297]: Just rethrow it, we should not use asserts.

File: bigtable-hbase-mirroring-client-1.x-parent/bigtable-hbase-mirroring-client-1.x/src/main/java/com/google/cloud/bigtable/mirroring/hbase1_x/MirroringConnection.java [hunk 2]
```diff
@@ -277,12 +283,36 @@ public Connection getSecondaryConnection() {
...
+    } catch (InterruptedException e) {
+      throw new IOException(e);
+    } catch (TimeoutException e) {
+      Log.trace("close() timeout");
```
Comment (mwalkiewicz) [inline bigtable-hbase-mirroring-client-1.x-parent/bigtable-hbase-mirroring-client-1.x/src/main/java/com/google/cloud/bigtable/mirroring/hbase1_x/MirroringConnection.java:301]: `.error`?

**Review by prawilny — COMMENTED** [review]
:lgtm: (with f2f nits)

> Reviewed 2 of 4 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/unoperate/java-bigtable-hbase-1/133#-:-MpRf5iJAFn-jhlyftUM:b-6360rc)* status: 2 of 4 files reviewed, 8 unresolved discussions (waiting on @mwalkiewicz and @prawilny)

**Review by mwalkiewicz — DISMISSED** [review]

**Review by mwalkiewicz — APPROVED** [review]

**Pull request merged**